File: 백준/Silver/18352. 특정 거리의 도시 찾기/특정 거리의 도시 찾기.py
from collections import deque
import sys
read = sys.stdin.readline
n, m, k, x = map(int, read().split())
l = [[] for _ in range(n + 1)]
visited = [-1] * (n + 1)
answer = []

for _ in range(m):
    a, b = map(int, read().split())
    l[a].append(b)

def BFS(v):
    q = deque()
    q.append(v)
    visited[v] += 1
    while q:
        p = q.popleft()
        for i in l[p]:
            if visited[i] == -1:
                visited[i] = visited[p] + 1
                q.append(i)

BFS(x)

for i in range(n + 1):
    if visited[i] == k:
        answer.append(i)

if not answer:
    print(-1)
else:
    answer.sort()
    for i in answer:
        print(i)

File: 백준/Silver/18352. 특정 거리의 도시 찾기/특정 거리의 도시 찾기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        StringBuilder sb = new StringBuilder();

        int N = Integer.parseInt(st.nextToken());
        int M = Integer.parseInt(st.nextToken());
        int K = Integer.parseInt(st.nextToken());
        int X = Integer.parseInt(st.nextToken());

        ArrayList<Integer>[] city = new ArrayList[N + 1];

        for (int i = 1; i <= N; i++) {
            city[i] = new ArrayList();
        }

        for (int i = 0; i < M; i++) {
            st = new StringTokenizer(br.readLine());

            int A = Integer.parseInt(st.nextToken());
            int B = Integer.parseInt(st.nextToken());

            city[A].add(B);
        }

        ArrayList<Integer> answer = new ArrayList<>();

        Queue<Integer> q = new ArrayDeque<>();
        boolean[] visited = new boolean[N + 1];
        q.offer(X);
        visited[X] = true;

        int dist = 1;

        while (!q.isEmpty()) {
            if (dist > K) {
                break;
            }

            int size = q.size();

            while (size-- > 0) {
                int cur = q.poll();

                for (Integer next : city[cur]) {
                    if (!visited[next]) {
                        if (dist == K) {
                            answer.add(next);
                        }

                        q.offer(next);
                        visited[next] = true;
                    }
                }
            }

            dist++;
        }

        Collections.sort(answer);

        if (answer.isEmpty()) {
            sb.append("-1");
        } else {
            for (Integer num : answer) {
                sb.append(num).append("\n");
            }
        }

        System.out.println(sb);
    }
}